src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -1285,21 +1285,149 @@ internal bool HasCollectionExpressionApplicableAddMethod(SyntaxNode syntax, Type
+                {
+                    return false;
+                }
+
+                // If the method is generic, fail if there are type arguments that are not referenced
+                // in the required parameters types. That is, fail if we know the type arguments cannot
+                // be inferred from the arguments.
+                var typeParameters = method.TypeParameters;
+                if (typeParameters.Length > 0)
+                {
+                    var requiredParameters = parameters.AsSpan(0, requiredLength);
+                    foreach (var typeParameter in typeParameters)
+                    {
+                        if (requiredParameters.All(typeParameter, (parameter, typeParameter) => !parameter.Type.ContainsTypeParameter(typeParameter)))
+                        {
+                            // The type parameter does not appear in any of the parameter types.
+                            return false;
+                        }
+                    }
+                }
+
+                return true;
+            }
+        }
+
         /// <summary>
         /// If the element is from a collection type where elements are added with collection initializers,
         /// return the argument to the collection initializer Add method or null if the element is not a
         /// collection initializer node. Otherwise, return the element as is.
         /// </summary>
-        internal static BoundExpression? GetUnderlyingCollectionExpressionElement(BoundCollectionExpression expr, BoundExpression? element)
+        internal static BoundExpression GetUnderlyingCollectionExpressionElement(BoundCollectionExpression expr, BoundExpression element, bool throwOnErrors)
         {
             if (expr.CollectionTypeKind is CollectionExpressionTypeKind.ImplementsIEnumerable)
             {
-                return element switch
+                switch (element)
                 {
-                    BoundCollectionElementInitializer collectionInitializer => getCollectionInitializerElement(collectionInitializer),
-                    BoundDynamicCollectionElementInitializer dynamicInitializer => dynamicInitializer.Arguments[0],
-                    _ => null,
-                };
+                    case BoundCollectionElementInitializer collectionInitializer:
+                        return getCollectionInitializerElement(collectionInitializer);
+                    case BoundDynamicCollectionElementInitializer dynamicInitializer:
+                        return dynamicInitializer.Arguments[0];
+                }
+                if (throwOnErrors)
+                {
+                    throw ExceptionUtilities.UnexpectedValue(element);
+                }
+                switch (element)
+                {
+                    case BoundCall call:
+                        return call.Arguments[call.InvokedAsExtensionMethod ? 1 : 0];
+                    default:
+                        return element;
```

Comment:
>return element; [](http://example.com/codeflow?start=24&length=15)

I assume this code is reachable. Therefore, the error recovery logic feels at least inaccurate. For example, it feels like the previous case can unwrap a call that actually was an underlying collection expression element.

Reply:
My expectation is the `BoundCall` case above is only used for `Add` calls, and all other error cases should be `BoundBadExpression` only. I've added an assert for `BoundBadExpression` to this case.